### PR TITLE
Implemented ScalarDamageBase

### LINKED ISSUE #11937: Provide ScalarDamageBase class
## Rationale

Scalar damage models (such as those implemented/to be implemented in Blackbear) use similar core parameters and methods. However, there is no unified ScalarDamageBase class that would unify these parameters. This leads to several chunks of code being repeated unnecessarily.

## Description

In the `ScalarMaterialDamage` class the following methods and attributes should be shared by a parent class: `updateStressForDamage` `updateJacobianMultiForDamage` `computeTimeStepLimit` `use_old_damage` `damage_index` `damage_index_old` `residual_stifdness_fraction` and `maximum_damage_increment`.

## Impact

Improved easiness to implement future scalar damage models.

### PR DESCRIPTION
**Pull request by agiorla**
Added a `ScalarDamageBase` class that is inherited by `ScalarMaterialDamage` so that its code can be reused.

Added a `CombinedScalarDamage` class that allows to compute a scalar damage from multiple `ScalarDamageBase` models.

Closes #11937

### CODE REVIEW

**Comment by agiorla** [comment]
@bwspenc I am not sure if we should leave the check in `ScalarDamageBase::updateDamage()` (as it is right now) or in `ScalarMaterialDamage::updateQpDamageIndex()`. Maybe we want to avoid doing too many checks for actual damage models such as the Mazars damage model.

**Comment by agiorla** [comment]
Tagging @bwspenc

**Comment by permcody** [comment]
This looks ok to me. Commit history is messing though 😉

File: modules/tensor_mechanics/include/materials/CombinedScalarDamage.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  void initialSetup() override;
+
+protected:
+  virtual void updateQpDamageIndex() override;
+
+  enum class CombinationType
+  {
+    Maximum,
+    Multiply
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/CombinedScalarDamage.h]: Would 'Product' be more appropriate?

File: modules/tensor_mechanics/src/materials/CombinedScalarDamage.C [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+validParams<CombinedScalarDamage>()
+{
+  InputParameters params = validParams<ScalarDamageBase>();
+
+  params.addClassDescription(
+      "Scalar damage model which is computed as a function of multiple scalar damage models");
+
+  params.addRequiredParam<std::vector<MaterialName>>("damage_models",
+                                                     "Name of the damage models used to compute "
+                                                     "the damage index ");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/CombinedScalarDamage.C]: Drop trailing space after 'index'

**Review by bwspenc — CHANGES_REQUESTED** [review]
Overall, this looks good to me.  Sorry for taking so long to get to this!  I just saw a couple of minor things to address.  If you could take care of those and squash your commits while you are at it, that would be great!

**Comment by agiorla** [comment]
@bwspenc No problem I know summer can be a busy time. I tried to squash the commits but I'm not sure it really did something...

File: modules/tensor_mechanics/doc/content/source/materials/CombinedScalarDamage.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+\begin{equation}
+    \boldsymbol{\sigma} = (1 - d)\ \mathbb{C} : \boldsymbol{\varepsilon}
+\end{equation}
+
+The damage variable $d$ itself is computed as a combination of a series of external scalar damage models defined by the `damage_models` input parameters.
+Two combination types are possible: `Maximum` (default) and `Product`:
+
+\begin{eqnarray}
+    \mathrm{Maximum:} & d = & \mathrm{max}(d_1 ... d_N) \\
+    \mathrm{Multiply:} & d = & 1 - \Prod\limits_{i=1}^{N} (1 - d_i)
```
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/materials/CombinedScalarDamage.md]: Multiply->Product

**Comment by bwspenc** [comment]
Just noticed one spot where you missed the name change. To rebase, do something like this:
```
git rebase -i HEAD~12
```
where `HEAD~12` is a commit before your full commit history.  That will put you in an editor, and you'll need to change all of the 'pick' words to 's' after your first commit in the series.  That will squash all of the commits marked 's' into the last one in the list marked 'pick'. You'll have a chance to edit the combined commit message.

It's not strictly necessary to do that, but it's really good to not clutter up the MOOSE commit history with a bunch of work in progress commits. Sometimes it makes sense to break a big change set into multiple commits for various reasons, but they should all be in a clean state.

**Comment by agiorla** [comment]
@bwspenc Thanks for the help. I had to do the rebase in sudo mode otherwise files were not overwritten properly for some reason, but it seems it worked.

**Review by bwspenc — APPROVED** [review]

**Comment by agiorla** [comment]
Thanks @bwspenc

**Pull request merged**